### PR TITLE
8321192 : j.a.PrintJob/ImageTest/ImageTest.java: Fail or skip the test if there's no printer

### PR DESCRIPTION
**Pull request by Renjithkannath**
Hi Reviewers,
I have updated the test with 'PassFailJFrame' with programmatically generating image and print , please review and let me know your suggestions.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8321192](https://bugs.openjdk.org/browse/JDK-8321192): j.a.PrintJob/ImageTest/ImageTest.java: Fail or skip the test if there's no printer (**Bug** - P4)

### Reviewers
 * [Alexey Ivanov](https://openjdk.org/census#aivanov) (@aivanov-jdk - **Reviewer**)
 * [Tejesh R](https://openjdk.org/census#tr) (@TejeshR13 - Committer) ⚠️ Review applies to [351bd721](https://git.openjdk.org/jdk/pull/17790/files/351bd721cb051a01ba9f78160007f7b16f5042d7)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17790/head:pull/17790` \
`$ git checkout pull/17790`

Update a local copy of the PR: \
`$ git checkout pull/17790` \
`$ git pull https://git.openjdk.org/jdk.git pull/17790/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17790`

View PR using the GUI difftool: \
`$ git pr show -t 17790`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17790.diff">https://git.openjdk.org/jdk/pull/17790.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17790#issuecomment-1936089643)

### CODE REVIEW

File: test/jdk/java/awt/PrintJob/ImageTest/ImageTest.java [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+import javax.imageio.ImageIO;
+import java.awt.Button;
```
Comment (aivanov-jdk) [inline test/jdk/java/awt/PrintJob/ImageTest/ImageTest.java:25]: Usually, `javax.*` packages are below `java.*` packages.

File: test/jdk/java/awt/PrintJob/ImageTest/ImageTest.java [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+                    width = dim.width - 30; // take care of paper margin
+                }
+                if (height > dim.height) {
+                    height = dim.height - 30;
+                }
+            }
+            g.drawImage(img, 10, 75, width, height, this);
+        }
+
+        public void setPrintJob(PrintJob pj) {
```
Comment (aivanov-jdk) [inline test/jdk/java/awt/PrintJob/ImageTest/ImageTest.java:78]: Can be make `private`.

File: test/jdk/java/awt/PrintJob/ImageTest/ImageTest.java [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+            if (pj != null) {
+                f.setPrintJob(pj);
+                Graphics pg = pj.getGraphics();
+                f.paint(pg);
+                pg.dispose();
+                pj.end();
+            }
+        });
+        f.add(b);
+        f.setBounds(0, 50, 700, 350);
```
Comment (aivanov-jdk) [inline test/jdk/java/awt/PrintJob/ImageTest/ImageTest.java:108]:
```suggestion
        f.setSize(700, 350);
```
The position is determined by `PassFailJFrame`.

**Review by aivanov-jdk — APPROVED** [review]
Should we mark `paint` and `imageUpdate` with `@Override` annotation?

Overall, looks good to me.

File: test/jdk/java/awt/PrintJob/ImageTest/ImageTest.java [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+             Click the Print button on the Frame. Select a printer from the
+             print dialog and click 'OK'. Verify that the image displayed
+             in the Frame is correctly printed. Test printing in both Color
+             and Monochrome.
+            """;
+
+    public static void main(String[] args) throws Exception {
+        if (PrinterJob.lookupPrintServices().length == 0) {
+            throw new SkippedException("Printer not configured or available."
+                    + " Test cannot continue.");
```
Comment (aivanov-jdk) [inline test/jdk/java/awt/PrintJob/ImageTest/ImageTest.java:151]:
```suggestion
            throw new SkippedException("Printer not configured or available.");
```
We decided to shorten the message. Let's keep it consistent.

**Comment by prrace** [comment]
I thought SkippedException looks like a "pass" too ?
What is effectively different by using it here ?
If the @printer keyword is there, then let's educate people that there MUST be a printer by making the test fail or if possible report a jtreg "error".  Not sure if that is possible, error is usually timeout and things like that.

**Comment by aivanov-jdk** [comment]
> I thought SkippedException looks like a "pass" too ?  
> What is effectively different by using it here ?

No, it's not. If a test throws `jtreg.SkippedException`, then it's reported as **skipped**. It is neither a failure nor a success — it indicates the prerequisites aren't met.

> If the @printer keyword is there, then let's educate people that there MUST be a printer by making the test fail or if possible report a jtreg "error". Not sure if that is possible, error is usually timeout and things like that.

That's exactly what `jtreg.SkippedException` does: it's a configuration / environment error.

Most printer tests already have such a condition. Previously, each test behaved as if the test passed even though it wasn't even run.

The `jtreg.SkippedException` changes the status which comes out as a distinct one from either *pass* or *fail*.

Do you think this check should be removed completely? Yes, it is somewhat redundant: the test with `@key printer` should not be started in an environment where there's no printer.

**Comment by aivanov-jdk** [comment]
> > I thought SkippedException looks like a "pass" too ?
> > What is effectively different by using it here ?
> 
> No, it's not. If a test throws `jtreg.SkippedException`, then it's reported as **skipped**. It is neither a failure nor a success — it indicates the prerequisites aren't met.

**jtreg FAQ: [What if a test does not apply in a given situation?](https://openjdk.org/jtreg/faq.html#what-if-a-test-does-not-apply-in-a-given-situation)**

“While `@modules` and `@requires` will cover many cases, there may be cases where the determination of whether the test is applicable needs to be determined by the test itself. For example, a test may want to check the presence of a shared library containing compiled native code, and to skip the main body of the test if the library is not available. <em>In such cases, a test may throw an exception named `jtreg.SkippedException`. […] This exception will be treated specially: unlike for any other exception, the test will be deemed to have passed, but the reason will be set to a message saying that the test was skipped instead of executing normally.</em>”

@prrace Does it explain `jtreg.SkippedException`?

However, `@key printer` also covers the condition in this case.

**Comment by prrace** [comment]
> > > I thought SkippedException looks like a "pass" too ?
> > > What is effectively different by using it here ?
> > 
> > 
> > No, it's not. If a test throws `jtreg.SkippedException`, then it's reported as **skipped**. It is neither a failure nor a success — it indicates the prerequisites aren't met.
> 
> **jtreg FAQ: [What if a test does not apply in a given situation?](https://openjdk.org/jtreg/faq.html#what-if-a-test-does-not-apply-in-a-given-situation)**
> 
> “While `@modules` and `@requires` will cover many cases, there may be cases where the determination of whether the test is applicable needs to be determined by the test itself. For example, a test may want to check the presence of a shared library containing compiled native code, and to skip the main body of the test if the library is not available. _In such cases, a test may throw an exception named `jtreg.SkippedException`. […] This exception will be treated specially: unlike for any other exception, the test will be deemed to have passed, but the reason will be set to a message saying that the test was skipped instead of executing normally._”
> 
> @prrace Does it explain `jtreg.SkippedException`?
> 
> However, `@key printer` also covers the condition in this case.

"the test will be deemed to have passed" is not what we want.
We want it to fail. So don't run these tests unless you have a printer. Otherwise we might as well leave the test as it is.

I see no point in using SkippedException in any of the printer tests unless there's some test-specific circumstance I can't think of.

**Comment by prrace** [comment]
> However, @key printer also covers the condition in this case.

Yes, if we didn't have a keyword, then it would be less reasonable to fail the test.
Like if a test requires multimon, and we don't have a keyword to support filtering such tests, THEN SkippedException might be reasonable.

Also adding that exception means you can't run without the jtreg framework, unless you provide the class as part of the test itself. I mean something has to provide that class.

**Comment by prrace** [comment]
I had a slack chat with Jon Gibbons (maintainer of jtreg) and he thinks FAIL would be better here.
I think we should be consistent and now that all tests that need a printer specify @key printer - 
should FAIL if there's no printer. Not saying we have to rush off and change them all immediately,
but if we are updating it anyway for some reason then do it.

**Comment by aivanov-jdk** [comment]
@prrace You're right. I was under impression that `jtreg.SkippedException` gives the test a distinct status from either *Pass* or *Fail*. Eventually, it's not the case.

In verbose mode, I can see the exception:

```text
runner starting test: java/awt/PrintJob/test/FailedTest.java
runner finished test: java/awt/PrintJob/test/FailedTest.java
Failed. Execution failed: `main' threw exception: java.lang.Error: Fail the test
runner starting test: java/awt/PrintJob/test/PassedTest.java
runner finished test: java/awt/PrintJob/test/PassedTest.java
Passed. Execution successful
runner starting test: java/awt/PrintJob/test/SkippedTest.java
runner finished test: java/awt/PrintJob/test/SkippedTest.java
Passed. Skipped: jtreg.SkippedException: Skip the test
Test results: passed: 2; failed: 1
```
This is why I thought it was recorded as a different status.

At the same time, the report looks like this:
```text
==============================
Test summary
==============================
   TEST                                              TOTAL  PASS  FAIL ERROR   
>> jtreg:open/test/jdk/java/awt/PrintJob/test            3     2     1     0 <<
==============================
TEST FAILURE
```

If `jtreg.SkippedException` resulted in a status distinct from *Pass* or *Fail*, it would be more useful.

**Comment by aivanov-jdk** [comment]
What would be the best option to handle a missing printer?

1. Throw a `RuntimeException` or `Error` indicating there's no printer available?
2. Let the test fail with whatever error it may fail when there's no printer?

I'm inclined towards the first option — it gives a clear indication of the error.

File: test/jdk/java/awt/PrintJob/ImageTest/ImageTest.java (hunk 4 above)
Comment (aivanov-jdk) [inline test/jdk/java/awt/PrintJob/ImageTest/ImageTest.java:151]: As per the above discussion, fail the test if there's no printer.

```suggestion
        if (PrinterJob.lookupPrintServices().length == 0) {
            throw new RuntimeException("Printer not configured or available.");
```

**Review by aivanov-jdk — CHANGES_REQUESTED** [review]

File: test/jdk/java/awt/PrintJob/ImageTest/ImageTest.java [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+import java.awt.print.PrinterJob;
+import java.io.File;
+import java.io.IOException;
+import javax.imageio.ImageIO;
+
+/*
+ * @test
+ * @bug 4242308 4255603
+ * @key printer
+ * @library /java/awt/regtesthelpers /test/lib
```
Comment (aivanov-jdk) [inline test/jdk/java/awt/PrintJob/ImageTest/ImageTest.java:44]:
```suggestion
 * @library /java/awt/regtesthelpers
```
The `/test/lib` library isn't used any more.

**Review by aivanov-jdk — APPROVED** [review]
Looks good except for minor comment: remove `/test/lib` from the libraries.

**Review by TejeshR13 — APPROVED** [review]
Looks good to me.

File: test/jdk/java/awt/PrintJob/ImageTest/ImageTest.java (hunk 5 above)
Comment (aivanov-jdk) [reply test/jdk/java/awt/PrintJob/ImageTest/ImageTest.java:44]: @Renjithkannath Please remove `/test/lib` from the list of libraries before integrating.

**Review by aivanov-jdk — APPROVED** [review]

**Pull request closed without merging**